### DISCUSSION TITLE
How CUTLASS deal with the last mma tile

### DISCUSSION THREAD

**Discussion by zihaozou** [NVIDIA/cutlass#591] [Q&A] [answered]
Hi,
I am studying the code and trying the locate the part that deals with the last MMA tile in the last "gemm_k_iterations". Since the last tile might be irregular to the predefined "InstructionShape"(in my case it's 16x8x8), how does the CUTLASS modulate this issue? 
Will it fill the empty part of the fragment with zeros?

**Comment by hwu36** [comment, marked as answer]
out of bound data will be padded with zero in the shared memory so nothing special is needed when calling tensor cores.  in sm80, we just need to use cp.async_zfill which can fill the shared memory with zeros automatically.

**Reply by zihaozou** [reply]
I can see that the "implicit_gemm_multistage" utilizes this function.

What about sm75 and set the kStages=2, which will use "implicit_gemm_pipelined"?

when loading data from global to shmem in "implicit_gemm_pipelined", it calls
`this->smem_iterator_A_.store`, which leads to `store_with_byte_offset` in ["regular_tile_iterator_tensor_op.h"](https://github.com/NVIDIA/cutlass/blob/1617685a77acba129dcbc12ab5d9c20ae230ae69/include/cutlass/transform/threadblock/regular_tile_iterator_tensor_op.h#L612).
In this part, I didn't see any code related to zero fill. Did I miss any important parts?

**Reply by zihaozou** [reply]
oh, it's in the [load](https://github.com/NVIDIA/cutlass/blob/1617685a77acba129dcbc12ab5d9c20ae230ae69/include/cutlass/conv/threadblock/conv2d_tile_iterator.h#L283).
Thank you.

**Reply by kris1025** [reply]
> out of bound data will be padded with zero in the shared memory so nothing special is needed when calling tensor cores. in sm80, we just need to use cp.async_zfill which can fill the shared memory with zeros automatically.

@hwu36  in sm75 mma single stage, how does cutlass ensure the padding memory is filled with zeros. the code is in https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/mma_singlestage.h ,the line 247 and 248.

**Reply by hwu36** [reply]
https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/mma_singlestage.h#L254-L255

**Reply by kris1025** [reply]
> https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/mma_singlestage.h#L254-L255

L245 and L255 is used to avoid out of bound memory access, not fill the frag to zeros.

**Reply by hwu36** [reply]
mask is off, then 0s are load.

**Reply by kris1025** [reply]
> mask is off, then 0s are load.

could you show me the code?

**Reply by hwu36** [reply]
https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/transform/threadblock/predicated_tile_iterator.h#L378-L379

**Reply by hwu36** [reply]
btw, the loading in the prologue can be smaller than the full threadblock tile.  mainloop always loads full threadblock tiles.  the last out of bound load is never loaded into the register file which means it is not used anyway.

**Reply by kris1025** [reply]
> https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/transform/threadblock/predicated_tile_iterator.h#L378-L379

this is in the store func. I don't understand why this function is called. Could you explain that?

**Reply by hwu36** [reply]
https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/transform/threadblock/predicated_tile_iterator.h#L339-L342 

load is just a few lines above.

**Reply by kris1025** [reply]
> https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/transform/threadblock/predicated_tile_iterator.h#L339-L342
> 
> load is just a few lines above.
Yes, for the last gemm iteration, the fragment is partially assigned by the global memory, ths other part will be zeros? this is the point i am wondering, because i don't see any code to deal with the other part.

**Reply by hwu36** [reply]
if you find anything wrong, you can try to change number 2 to 1 in https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/mma_singlestage.h#L254-L255.  This condition looks wrong to be.

**Reply by kris1025** [reply]
> if you find anything wrong, you can try to change number 2 to 1 in https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/mma_singlestage.h#L254-L255. This condition looks wrong to be.

https://github.com/NVIDIA/cutlass/blob/1617685a77acba129dcbc12ab5d9c20ae230ae69/include/cutlass/conv/threadblock/conv2d_tile_iterator.h#L283 , in the conv2d tile iterator , it always clean the fragment first to ensure the fragment is filled with zeros.  Is there any bug in predicted tile iterator?

**Comment by kris1025** [comment]
您好！您发送给我的邮件已收到。华林泉
